ffi/derive/src/convert.rs

```diff
@@ -44,449 +45,763 @@ pub fn derive_into_ffi(input: &DeriveInput) -> TokenStream2 {
+        });
+
+        let (discriminants, discriminant_decls) = gen_discriminants(enum_name, enum_, &tag_type);
+
+        quote! {
+            #repr_c_enum
 
-            fn into_ffi(self) -> Self::Target {
-                let layout = core::alloc::Layout::for_value(&self);
+            impl<'itm> iroha_ffi::TryFromReprC<'itm> for #enum_name {
+                type Source = #repr_c_enum_name<'itm>;
+                type Store = (#(#store),*);
 
-                unsafe {
-                    let ptr: Self::Target = alloc(layout).cast();
-                    ptr.write(self);
-                    ptr
+                unsafe fn try_from_repr_c(
+                    source: <Self as iroha_ffi::TryFromReprC<'itm>>::Source,
+                    store: &'itm mut <Self as iroha_ffi::TryFromReprC<'itm>>::Store
+                ) -> iroha_ffi::Result<Self> {
+                    #( #discriminant_decls )*
+
+                    match source.tag {
+                        #( #discriminants => Ok(#enum_variants), )*
+                        _ => Err(iroha_ffi::FfiReturn::TrapRepresentation),
+                    }
                 }
             }
         }
+    }
 
-        impl iroha_ffi::IntoFfi for &#name {
-            type Target = *const #name;
-
-            fn into_ffi(self) -> Self::Target {
-                <*const _>::from(self)
+    pub fn derive_into_ffi_for_data_carrying_enum(
+        enum_name: &Ident,
+        enum_: &syn::DataEnum,
+    ) -> TokenStream2 {
+        let (repr_c_enum_name, repr_c_enum) = gen_repr_c_enum(enum_name, enum_, true);
+
+        let enum_variants = enum_.variants.iter().enumerate().map(|(i, variant)| {
+            let idx = TokenStream2::from_str(&format!("{i}")).expect("Valid");
+            let variant_name = &variant.ident;
+
+            match &variant.fields {
+                syn::Fields::Unnamed(syn::FieldsUnnamed { unnamed, .. }) if unnamed.len() == 1 => {
+                    let payload_name = gen_repr_c_enum_payload_name(enum_name, true);
+
+                    quote! {
+                        #enum_name::#variant_name(x) => {
+                            let payload = #payload_name {
+                                #variant_name: core::mem::ManuallyDrop::new(
+                                    iroha_ffi::IntoFfi::into_ffi(x)
+                                )
+                            };
+
+                            #repr_c_enum_name { tag: #idx, payload }
+                        },
+                    }
+                }
+                syn::Fields::Unnamed(syn::FieldsUnnamed { unnamed, .. }) => {
+                    abort!(unnamed, "Only 1-sized variants are supported")
+                }
+                syn::Fields::Named(syn::FieldsNamed { named, .. }) => {
+                    abort!(named, "Named variants are not supported")
+                }
+                syn::Fields::Unit => {
+                    quote! { #enum_name::#variant_name => #repr_c_enum_name { tag: #idx }, }
```

Comment:
Another approach could be to store `tag` inside union in the same place for every variant ([as proposed here](https://github.com/rust-lang/rfcs/pull/2195)), this could lead to more compact memory layout.

Reply:
Prefer the other approach. This is less surprising to people who are only familiar with Rust.